File: compatibility-validator/app/static/cabling.js
/* Installation metadata and revision-bound cabling documents for the project. */
(() => {
  'use strict';
  const { esc, RequestGate, statusClass } = ValidatorCore;
  const $ = id => document.getElementById(id);
  const gate = new RequestGate();
  let context, report = null;
  const empty = () => ({ locations: [], port_labels: [], cables: [] });
  const entries = project => [...project.connections, ...project.breakouts];
  const selections = project => [...project.connections.flatMap(c => [c.a, c.b]), ...project.breakouts.flatMap(b => [b.head, ...b.branches.map(v => v.selection)])];
  const portKey = p => JSON.stringify([p.instance_id, p.port_group_id, p.port_number]);
  const endpointText = p => `${p.rack || 'Rack unknown'}${p.rack_u ? ' / U' + p.rack_u : ''}\n${p.instance_id}\n${p.port_label || 'Label unconfirmed'} (${p.port_group_id}/${p.port_number})${p.optical_port ? ' / optical ' + p.optical_port : ''}`;
  const field = (attributes, value, label, type = 'text') => `<input ${attributes} type="${type}" aria-label="${esc(label)}" value="${esc(value ?? '')}" maxlength="64"${type === 'number' ? ' min="1" max="1000" step="1"' : ''}>`;

  function options(project) { project.cabling ||= empty(); return project.cabling; }
  function cable(project, id) {
    const values = options(project).cables;
    let value = values.find(c => c.entry_id === id);
    if (!value) { value = { entry_id: id, cable_id: null, progress: [] }; values.push(value); }
    return value;
  }
  function invalidate() {
    gate.cancel(); report = null; $('cablingReport').replaceChildren();
    $('cablingStatus').textContent = 'Build the plan to review connections and export current documents.';
    for (const id of ['downloadCablingPDF', 'downloadLabelsPDF', 'downloadCablingXLSX']) $(id).disabled = true;
    $('buildCabling').disabled = false;
  }
  function render(project) {
    const data = project.cabling || empty(), all = selections(project);
    const instances = [...new Set(all.map(p => p.instance_id))].sort();
    const ports = [...new Map(all.map(p => [portKey(p), p])).values()];
    $('cablingLocations').innerHTML = instances.length ? `<table><thead><tr><th>Physical device</th><th>Rack</th><th>Rack unit (optional)</th></tr></thead><tbody>${instances.map(id => {
      const value = data.locations.find(p => p.instance_id === id) || {};
      return `<tr><td>${esc(id)}</td><td>${field(`data-location="${esc(id)}" data-field="rack"`, value.rack, id + ' rack')}</td><td>${field(`data-location="${esc(id)}" data-field="rack_u"`, value.rack_u, id + ' rack unit', 'number')}</td></tr>`;
    }).join('')}</tbody></table>` : '<p>Add a connection or breakout to prepare its installation plan.</p>';
    $('cablingPorts').innerHTML = `<table><thead><tr><th>Device / physical cage</th><th>Label printed on device</th></tr></thead><tbody>${ports.map((p, index) => {
      const value = data.port_labels.find(v => portKey(v) === portKey(p));
      return `<tr><td>${esc(p.instance_id)} / ${esc(p.port_group_id)} / ${esc(p.port_number)}</td><td>${field(`data-port-index="${index}"`, value?.label, p.instance_id + ' physical port ' + p.port_number)}</td></tr>`;
    }).join('')}</tbody></table>`;
    $('cablingIds').innerHTML = `<table><thead><tr><th>Project entry</th><th>Unique cable ID (blank uses C-entry)</th></tr></thead><tbody>${entries(project).map(e => `<tr><td>${esc(e.id)}</td><td>${field(`data-cable="${esc(e.id)}" placeholder="${esc('C-' + e.id)}"`, data.cables.find(c => c.entry_id === e.id)?.cable_id, e.id + ' cable ID')}</td></tr>`).join('')}</tbody></table>`;
    $('buildCabling').disabled = !instances.length;
    $('cablingPorts').dataset.keys = JSON.stringify(ports.map(p => ({ instance_id: p.instance_id, port_group_id: p.port_group_id, port_number: p.port_number })));
  }
  function prune(project) {
    if (!project.cabling) return;
    const all = selections(project), ports = new Set(all.map(portKey)), instances = new Set(all.map(p => p.instance_id));
    project.cabling.locations = project.cabling.locations.filter(p => instances.has(p.instance_id));
    project.cabling.port_labels = project.cabling.port_labels.filter(p => ports.has(portKey(p)));
    project.cabling.cables = project.cabling.cables.filter(c => entries(project).some(e => e.id === c.entry_id));
  }
  function shape(project) {
    if (project.cabling === undefined) return;
    const c = project.cabling;
    if (!c || typeof c !== 'object' || !['locations', 'port_labels', 'cables'].every(k => Array.isArray(c[k]) && c[k].length <= 512 && c[k].every(v => v && typeof v === 'object'))) throw new Error('Cabling metadata needs locations, port_labels and cables arrays.');
    if (!c.cables.every(v => Array.isArray(v.progress) && v.progress.every(p => p && typeof p === 'object'))) throw new Error('Cable installation entries need progress arrays.');
  }
  function download(blob, filename) {
    const url = URL.createObjectURL(blob), a = document.createElement('a');
    a.href = url; a.download = filename; a.click(); setTimeout(() => URL.revokeObjectURL(url), 1000);
  }
  async function request(path, binary = false) {
    let payload;
    try { payload = context.getProject(); } catch (error) { $('cablingStatus').textContent = error.message; return; }
    const ticket = gate.start(context.key());
    const timer = setTimeout(() => ticket.controller.abort(), 30000);
    $('cablingStatus').textContent = binary ? 'Preparing document…' : 'Building the complete installation plan…';
    try {
      const response = await fetch(path, { method: 'POST', headers: { 'Content-Type': 'application/json' }, cache: 'no-store', signal: ticket.controller.signal, body: JSON.stringify(payload) });
      if (!response.ok) {
        const data = await response.json();
        throw new Error(Array.isArray(data.detail) ? data.detail.slice(0, 5).map(v => `${v.loc.join('.')}: ${v.msg}`).join('; ') : data.detail || 'Cabling request failed');
      }
      const value = binary ? await response.blob() : await response.json();
      if (!gate.current(ticket, context.key())) return;
      const revision = binary ? response.headers.get('X-Catalog-Revision') : value.revision;
      if (revision !== payload.revision) throw new Error('Catalog changed. Refresh and build the plan again.');
      if (binary) {
        download(value, path.endsWith('labels.pdf') ? 'cable-labels.pdf' : 'cabling-plan.' + (path.endsWith('xlsx') ? 'xlsx' : 'pdf'));
        $('cablingStatus').textContent = 'Document exported from the current project.';
      } else { report = value; renderReport(); }
    } catch (error) {
      if (ticket.sequence === gate.sequence) $('cablingStatus').textContent = error.name === 'AbortError' ? 'Request timed out. Try again.' : error.message;
    } finally { clearTimeout(timer); }
  }
  function renderReport() {
    const s = report.summary;
    $('cablingStatus').textContent = `${s.cables} cables · ${s.legs} legs · ${s.labels} end labels · installed ${s.installed}/${s.legs} · checked ${s.checked}/${s.legs}`;
    $('cablingReport').innerHTML = `<p><span class="pill ${statusClass(report.status)}">${esc(report.status)}</span> ${report.provisional ? 'Provisional plan: review compatibility and missing installation details.' : 'All required plan checks passed.'}</p><div class="tablewrap"><table><thead><tr><th>Cable / branch</th><th>Source</th><th>Destination</th><th>Cable PN / length</th><th>Installed</th><th>Checked</th></tr></thead><tbody>${report.rows.map((r, i) => `<tr><td><strong>${esc(r.cable_id)}</strong><div>${esc(r.entry_id)}${r.branch_id ? ' / ' + esc(r.branch_id) + ' (termination ' + esc(r.termination) + ')' : ''}</div>${r.progress_stale ? '<div class="warn">Previous confirmation is stale</div>' : ''}${r.notes ? `<div>${esc(r.notes)}</div>` : ''}</td><td class="endpoint-text">${esc(endpointText(r.source))}</td><td class="endpoint-text">${esc(endpointText(r.destination))}</td><td>${esc(r.part_number || 'PN unknown')}<div>${r.length_m == null ? 'Length unknown' : esc(r.length_m) + ' m'} · ${esc(r.fabric)}</div></td><td><input type="checkbox" data-progress="${i}" data-field="installed" aria-label="${esc(r.cable_id + ' ' + (r.branch_id || '') + ' installed')}"${r.installed ? ' checked' : ''}></td><td><input type="checkbox" data-progress="${i}" data-field="checked" aria-label="${esc(r.cable_id + ' ' + (r.branch_id || '') + ' checked')}"${r.checked ? ' checked' : ''}></td></tr>`).join('')}</tbody></table></div><details><summary>${report.issues.length} installation details to complete</summary><ul>${report.issues.map(v => `<li>${esc(v)}</li>`).join('')}</ul></details><details><summary>Preview end labels</summary><div class="label-preview">${report.labels.map(l => `<article><strong>${esc(l.label_id)}</strong><div class="endpoint-text">${esc(endpointText(l.local))}</div><p>Remote: ${l.peers.length === 1 ? esc(l.peers[0].instance_id + ' / ' + (l.peers[0].port_label || l.peers[0].port_group_id + '/' + l.peers[0].port_number)) : esc(l.peers.length + ' branches; see plan')}</p><span class="small">${esc(l.part_number || 'PN unknown')}</span></article>`).join('')}</div></details>`;
    for (const id of ['downloadCablingPDF', 'downloadLabelsPDF', 'downloadCablingXLSX']) $(id).disabled = false;
  }
  for (const id of ['cablingLocations', 'cablingPorts', 'cablingIds']) $(id).addEventListener('input', event => {
    const input = event.target, d = input.dataset;
    try {
      context.change(project => {
        const data = options(project);
        if (d.location !== undefined) {
          let value = data.locations.find(p => p.instance_id === d.location);
          if (!value) { value = { instance_id: d.location, rack: null, rack_u: null }; data.locations.push(value); }
          value[d.field] = d.field === 'rack_u' ? (input.value ? Number(input.value) : null) : input.value.trim() || null;
        } else if (d.portIndex !== undefined) {
          const port = JSON.parse($('cablingPorts').dataset.keys)[Number(d.portIndex)];
          data.port_labels = data.port_labels.filter(p => portKey(p) !== portKey(port));
          if (input.value.trim()) data.port_labels.push({ ...port, label: input.value.trim() });
        } else if (d.cable !== undefined) cable(project, d.cable).cable_id = input.value.trim() || null;
      });
    } catch (error) { $('cablingStatus').textContent = error.message; }
  });
  $('cablingReport').addEventListener('change', event => {
    const d = event.target.dataset;
    if (d.progress === undefined || !report) return;
    const row = report.rows[Number(d.progress)], checked = event.target.checked;
    try {
      context.change(project => {
        const record = cable(project, row.entry_id);
        let progress = record.progress.find(p => (p.branch_id || null) === row.branch_id);
        if (!progress) { progress = { branch_id: row.branch_id, notes: '' }; record.progress.push(progress); }
        Object.assign(progress, { installed: row.installed, checked: row.checked, definition: row.definition });
        progress[d.field] = checked;
        if (d.field === 'checked' && checked) progress.installed = true;
        if (d.field === 'installed' && !checked) progress.checked = false;
      });
      request('/api/project/cabling');
    } catch (error) { $('cablingStatus').textContent = error.message; }
  });
  $('buildCabling').addEventListener('click', () => request('/api/project/cabling'));
  for (const [id, path] of [['downloadCablingPDF', 'cabling.pdf'], ['downloadLabelsPDF', 'labels.pdf'], ['downloadCablingXLSX', 'cabling.xlsx']]) $(id).addEventListener('click', () => { if (report) request('/api/project/' + path, true); });
  window.addEventListener('pagehide', () => gate.cancel());
  window.ValidatorCabling = { init: value => { context = value; }, render, invalidate, prune, shape };
})();

File: compatibility-validator/app/cabling.py
"""Installation rows and one label per physical cable termination."""

import hashlib
import json


def build_cabling(snapshot, project, validation):
    products = {p["id"]: p for p in snapshot["interconnects"]}
    devices = {d["id"]: d for d in snapshot["devices"]}
    locations = {p.instance_id: p.model_dump(exclude={"instance_id"}) for p in project.cabling.locations}
    port_labels = {(p.instance_id, p.port_group_id, p.port_number): p.label for p in project.cabling.port_labels}
    settings = {c.entry_id: c for c in project.cabling.cables}
    validations = {r["id"]: r for r in validation["results"]}
    sku_lengths = {s["part_number"]: s.get("length_m") for p in snapshot["interconnects"] for s in p["skus"]}
    sku_lengths.update({s["part_number"]: s["length_m"] for s in snapshot.get("fiber_assemblies", [])})
    rows, labels = [], []

    def endpoint(selection, optical_port=None):
        device = devices.get(selection.device_id, {})
        item = products.get(selection.product_id, {})
        return {"instance_id": selection.instance_id, "model": device.get("model", selection.device_id),
            "port_group_id": selection.port_group_id, "port_number": selection.port_number,
            "port_label": port_labels.get((selection.instance_id, selection.port_group_id, selection.port_number)),
            "optical_port": optical_port, "mode_id": selection.mode_id,
            "product_pn": selection.part_number,
            "module_pn": selection.part_number if item.get("category") == "Transceiver" else None,
            "rack": locations.get(selection.instance_id, {}).get("rack"),
            "rack_u": locations.get(selection.instance_id, {}).get("rack_u")}

    def label(cable_id, suffix, local, peers, pn):
        value = {"label_id": cable_id + "/" + suffix, "cable_id": cable_id, "end": suffix,
                 "local": local, "peers": peers, "part_number": pn}
        labels.append(value)
        return value["label_id"]

    def add_row(entry, branch, cable_id, pn, a, b, a_label, b_label, progress):
        branch_id = branch.id if branch else None
        length = entry.length_m if entry.length_m is not None else sku_lengths.get(pn)
        config = entry.model_dump(exclude={"revision"})
        if "branches" in config:
            config["branches"].sort(key=lambda b: b["id"])
        # Whole-breakout changes invalidate its old leg declarations. Metadata and
        # status timestamps are deliberately not inferred from file import time.
        payload = {"entry": config, "branch": branch_id, "cable_id": cable_id,
                   "source": a, "destination": b, "part_number": pn, "length_m": length}
        fingerprint = hashlib.sha256(json.dumps(payload, sort_keys=True, ensure_ascii=True, separators=(",", ":")).encode()).hexdigest()
        stored = next((p for p in progress if p.branch_id == branch_id), None)
        stale = bool(stored and (stored.installed or stored.checked) and stored.definition != fingerprint)
        result = validations[entry.id]
        row = {"entry_id": entry.id, "branch_id": branch_id,
            "termination": branch.termination if branch else None,
            "head_links": branch.head_links if branch else [],
            "head_optical_lanes": branch.head_optical_lanes if branch else [],
            "branch_optical_lanes": branch.branch_optical_lanes if branch else [],
            "cable_id": cable_id, "part_number": pn,
            "length_m": length,
            "documented_length_m": sku_lengths.get(pn), "fabric": entry.fabric,
            "source": a, "destination": b, "source_label": a_label, "destination_label": b_label,
            "validation_status": result["status"], "definition": fingerprint,
            "installed": bool(stored and stored.installed and not stale),
            "checked": bool(stored and stored.checked and not stale), "progress_stale": stale,
            "notes": stored.notes if stored else ""}
        rows.append(row)

    for kind, entry in [("connection", e) for e in project.connections] + [("breakout", e) for e in project.breakouts]:
        setting = settings.get(entry.id)
        cable_id = setting.cable_id if setting and setting.cable_id else "C-" + entry.id
        progress = setting.progress if setting else []
        if kind == "connection":
            optical = any(products.get(s.product_id, {}).get("category") == "Transceiver" for s in (entry.a, entry.b))
            pn = entry.fiber_part_number if optical else entry.a.part_number
            a, b = endpoint(entry.a), endpoint(entry.b)
            al, bl = label(cable_id, "A", a, [b], pn), label(cable_id, "B", b, [a], pn)
            add_row(entry, None, cable_id, pn, a, b, al, bl, progress)
        else:
            optical = entry.topology == "optical"
            pn = entry.optical_fanout.part_number if optical and entry.optical_fanout else None if optical else entry.head.part_number
            heads = {}
            ports = set(b.head_optical_port if optical else 0 for b in entry.branches)
            if optical and entry.optical_fanout:
                ports.update(range(1, entry.optical_fanout.head_ports + 1))
            for port in sorted(ports):
                a = endpoint(entry.head, port or None)
                peers = [endpoint(b.selection) for b in entry.branches if not optical or b.head_optical_port == port]
                heads[port] = (a, label(cable_id, "H" + (str(port) if optical else ""), a, peers, pn))
            for branch in sorted(entry.branches, key=lambda b: (b.termination, b.id)):
                a, al = heads[branch.head_optical_port if optical else 0]
                b = endpoint(branch.selection)
                bl = label(cable_id, "BR-" + branch.id, b, [a], pn)
                add_row(entry, branch, cable_id, pn, a, b, al, bl, progress)
    missing = []
    for row in rows:
        for side in ("source", "destination"):
            end = row[side]
            if not end["rack"]:
                missing.append(f"{row['cable_id']}: rack missing for {end['instance_id']}.")
            if not end["port_label"]:
                missing.append(f"{row['cable_id']}: verify physical port label for {end['instance_id']}/{end['port_group_id']}/{end['port_number']}.")
        if not row["part_number"]:
            missing.append(f"{row['cable_id']}: cable/harness ordering PN is unknown.")
        if row["length_m"] is None:
            missing.append(f"{row['cable_id']}: actual cable/path length is unknown.")
    issues = list(dict.fromkeys(missing))
    return {"scope": "cabling-plan", "name": project.name, "revision": snapshot["revision"],
        "status": validation["status"], "provisional": validation["status"] != "compatible" or bool(issues),
        "rows": rows, "labels": labels, "issues": issues,
        "summary": {"cables": len(project.connections) + len(project.breakouts), "legs": len(rows), "labels": len(labels),
            "installed": sum(r["installed"] for r in rows), "checked": sum(r["checked"] for r in rows),
            "stale_confirmations": sum(r["progress_stale"] for r in rows)},
        "notes": ["Installed and checked are user declarations for this exact installation definition; they do not change compatibility or manufacturer qualification.",
            "A breakout is one assembly: rows describe its legs, and labels identify the shared head and each physical branch.",
            "Named port labels are user-supplied markings; the catalog validation still uses the physical group and 1-based cage ordinal.",
            "Project JSON preserves locations, cable IDs, port labels and installation progress. PDF and XLSX are report snapshots."]}


def endpoint_text(end, model=False):
    rack = end["rack"] or "Rack unknown"
    if end["rack_u"] is not None:
        rack += " / U" + str(end["rack_u"])
    port = end["port_label"] or "Label unconfirmed"
    port += f" ({end['port_group_id']}/{end['port_number']})"
    if end["optical_port"]:
        port += f" / optical {end['optical_port']}"
    parts = [rack, end["instance_id"], port]
    if model:
        parts += [end["model"], "Mode: " + (end["mode_id"] or "auto")]
        if end["module_pn"]:
            parts.append("Module PN: " + end["module_pn"])
    return "\n".join(parts)
